### PR TITLE
coinjoin: implement blame rounds

### LINKED ISSUE #7181: Implement CoinJoin blame rounds

### PR DESCRIPTION
**Pull request by szymonlesisz**
## Description

If currently signed round ends with `NotAllAlicesSign` exception detain used inputs and outputs to be used again in blame round.

Blame round is a specific round dedicated **only** for participants who correctly sign previously failed round.
Participants who didn't sign or new participants will not be accepted. (will receive `InputNotWhitelisted` error during input-registration) 

#### Additional changes
`arrayToDictionary` required some tweaks to make it useful in my case (optional keys)
while doing so i've also added:
- strongly typed keys feature
> before: the key was always `string` even if `number` was acceptable return type of `getKey` callback
> now: dictionary key type is deducted from the return type of `getKey` callback
- getKey result runtime validation

`arrayToDictionary` still needs some type fixes regarding accepted params ([1, 2, 3] as const)

- code hygiene commit, using `arrayToDictionary` in few coinjoin related actions

## Related Issue

Resolve #7181

### CODE REVIEW

File: packages/coinjoin/src/client/CoinjoinPrison.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ export interface PrisonInmate {
     sentenceEnd: number;
     reason?: string;
     roundId?: string;
+    blameOf?: string;
```
Comment (szymonlesisz) [inline packages/coinjoin/src/client/CoinjoinPrison.ts]: unused

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 2]
```diff
@@ -75,6 +79,26 @@ export const getAccountCandidates = (
 
         // TODO: double-check account max signed rounds, should be done by suite tho
 
+        const blameOfUtxos = account.utxos.reduce<Record<string, Account['utxos']>>((res, utxo) => {
+            const a = blameOfInputs.find(i => i.id === utxo.outpoint);
```
Comment (szymonlesisz) [inline packages/coinjoin/src/client/round/selectRound.ts]: rename params

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 3]
```diff
@@ -107,19 +132,68 @@ export const getAccountCandidates = (
...
+const selectInputsForBlameRound = (
+    generator: AliceGenerator,
+    roundCandidates: CoinjoinRound[],
+    accountCandidates: ReturnType<typeof getAccountCandidates>,
+    { log }: CoinjoinRoundOptions,
+) =>
+    roundCandidates.find(round => {
+        const inputs: CoinjoinRound['inputs'] = [];
+        accountCandidates.forEach(account => {
+            const utxos = account.blameOf ? account.blameOf[round.blameOf] : null;
```
Comment (szymonlesisz) [inline packages/coinjoin/src/client/round/selectRound.ts:144]:
```suggestion
            const utxos = account.blameOf ? account.blameOf[round.blameOf] : [];
```

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 4]
```diff
@@ -75,6 +79,26 @@ export const getAccountCandidates = (
 
         // TODO: double-check account max signed rounds, should be done by suite tho
 
+        const blameOfUtxos = account.utxos.reduce<Record<string, Account['utxos']>>((res, utxo) => {
```
Comment (szymonlesisz) [inline packages/coinjoin/src/client/round/selectRound.ts]: @marekrjpolak you've mentioned some utility for that?

Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts]: `arrayToDictionary` 👍 found it :)

Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts]: used and explained in description

File: packages/coinjoin/src/client/CoinjoinRound.ts [hunk 5]
```diff
@@ -225,12 +227,19 @@ export class CoinjoinRound extends EventEmitter {
                 );
 
                 this.addresses.forEach(addr =>
-                    prison.detain(addr.address, {
+                    prison.detain(addr.scriptPubKey, {
```
Comment (dahaca) [inline packages/coinjoin/src/client/CoinjoinRound.ts:230]: just curious, why this change?

File: packages/coinjoin/src/client/round/selectRound.ts [hunk 6]
```diff
@@ -75,6 +79,24 @@ export const getAccountCandidates = (
...
+            account.utxos,
+            utxo => {
+                const blamedUtxo = blameOfInputs.find(i => i.id === utxo.outpoint);
+                return blamedUtxo?.roundId;
+            },
+            true,
+        );
+
+        if (Object.keys(blameOfUtxos).length > 0) {
+            log(`Found account candidate for blame round ~~${accountKey}~~`);
```
Comment (dahaca) [inline packages/coinjoin/src/client/round/selectRound.ts:92]: It would be OK to treat blame rounds as normal rounds when it comes to `SessionPhases`, right?

File: packages/utils/src/arrayToDictionary.ts [hunk 7]
```diff
@@ -1,32 +1,50 @@
...
  * @param array Array to be converted to dictionary
  * @param getKey Function extracting string from an array item T, which will become its
  * key in the dictionary (if not unique, latter item could replace the former one)
+ * Item will not be added to dictionary if key is not defined
  * @param multiple If true, dictionary values are arrays of all items with the given key
  * @returns Dictionary object with array items as values
  */
-export const arrayToDictionary: ArrayToDictionary = <T>(
+
+const validateKey = (key: any): key is DictionaryKey => {
```
Comment (dahaca) [inline packages/utils/src/arrayToDictionary.ts]: nit: I think we should start using `unknown` more, for situations where a function actually _doesn't know_ the incoming type and not when it _doesn't care_ about the type. `validateKey()` does care about the type eventually since it does the type check and returns based on the outcome. Overall, `any` is not even a type but a way to bypass type-checking altogether as I see it :D With better type safety, `unknown` could help to catch possible errors early on. What do you say? 

Btw here it obviously would not affect anything, more of a stylistic choice but getting used to `unknown` in general could still be useful.

**Review by dahaca — APPROVED** [review]

File: packages/coinjoin/src/client/CoinjoinRound.ts (hunk 5 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/CoinjoinRound.ts:230]: output registration checks if [scriptPubKey is detained](https://github.com/trezor/trezor-suite/blob/develop/packages/coinjoin/src/client/round/outputRegistration.ts#L52)

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 6 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/round/selectRound.ts:92]: not my decision what should be displayed in UI, ask @hynek-jina

File: packages/utils/src/arrayToDictionary.ts (hunk 7 above)
Comment (szymonlesisz) [reply packages/utils/src/arrayToDictionary.ts]: i dont have a strong opinion on that, it that particular case it doesnt really matter.
changed to uknown, rebased and squashed

Comment (dahaca) [reply packages/utils/src/arrayToDictionary.ts]: Perfect! Added to the Code Style's [proposals section](https://www.notion.so/satoshilabs/Proposals-6296b830b8e94df889ba93f66f7964a5).

File: packages/coinjoin/src/client/round/selectRound.ts (hunk 6 above)
Comment (hynek-jina) [reply packages/coinjoin/src/client/round/selectRound.ts:92]: Yes, user doesn't need to understand the concept of blame rounds..
